blocks/monitoring/__init__.py

```diff
@@ -0,0 +1,193 @@
+"""This part of the framework helps you monitor your model during training.
+
+:class:`MonitorChannels` describe quantities that should be monitored. They
+can either be Theano variables (e.g. the objective function, the weight
+norms, etc.) or they can be callable functions (e.g. if you want to use
+NumPy to calculate eigenvalues, or sample text from your language model).
+
+.. todo::
+
+   Monitored Theano variables might depend on inputs beyond those given to
+   the model. There should be a way of providing these to the monitoring
+   channels, collecting them, and compiling them as part of the Theano
+   function that performs monitoring.
+
+"""
+import logging
+from abc import ABCMeta, abstractmethod
+
+from blocks.utils import shared_expression, update_instance
+
+logger = logging.getLogger(__name__)
+
+
+class MonitorChannel(object):
```

Comment:
Bart,


On 1/7/2015 5:16 PM, Bart van Merriënboer wrote:
> I didn't say that the monitoring extension needs to collect /all/ 
> monitoring channels. It can obviously choose a subset of them. A 
> /channel/ (or aggregator in your preferred wording) doesn't mean it 
> /is/ monitored, merely that it /can/ be. 
Ok, sorry. The misunderstanding is because for me everything can be 
monitored. The aggregation is in a sense just a workaround for memory 
limitations.
>
> You say that aggregators refer to how variables are monitored on large 
> datasets. The counter-example I mentioned before is the 
> |ModelProperty| class you wrote, which suggests that "aggregators" 
> aren't just used to monitor variables over multiple batches, but 
> /also/ to monitor other Theano variables which don't rely on data. The 
> |ModelProperty| doesn't add any useful information on *how* to the 
> variable should be monitored, the only thing it does it tagging a 
> variable in order to signify that it /can/ be monitored.
>
the ModelProperty is just a convenience class, such that evaluation code 
is simpler. Instead of having ifs, I created a dummy aggregation scheme 
that wraps variables that don't need aggregation. It is quite common 
practice in programming. It can be a completely hidden, internal to the 
evaluator class. All its purpose is to make writing evaluation code easy.

Let me give you a totally unrelated example of the practice. Say you 
look for the index of something in an array. You append what you look 
for at the end of the array. Then you are guaranted to find it and you 
don't have to have ifs to detect the end of the array (people would do 
it for speed). In the same spirit, in my selectors there is the 
TerminatorSelector 
https://github.com/janchorowski/blocks/blob/selectors/blocks/select.py#L297 
which raises parse errors.
>
> The way I see it we can do two things:
>
> 1.
>
>     We can /get rid off/ the |ModelProperty| class and treat variables
>     which don't aggregation entirely differently, in which case
>     |Aggregator| makes sense. In that case we need another way of
>     attaching variables that can be monitored but don't need data to
>     the application call.
>
So lets completely remove the ModelProperty form the public interface 
and keep it for internal use only. If something does not depend on data, 
then there is no need to force the user to provide an aggregation 
scheme. If the evaluator is implemented like I did it, the ModelProperty 
can be a thing totaly internal to the evaluator, just a programming 
convenience.
>
> 1.
>
>     The other solution is the one I tried to convey here: A single
>     class (|VariableMonitorChannel|) is used to designate that a
>     Theano variable is monitorable (not monitored!). These channels
>     have "optional" aggregation schemes, the |ModelProperty| class
>     being used for variables which don't need any. In this case I feel
>     that we need to avoid the word |Aggregator| because for a large
>     subset nothing is actually aggregated.
>
I prefer to keep the aggregators to say how to evaluate variables 
separate from the concept of monitoring.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/bartvm/blocks/pull/90/files#r22595636>.
>